### ISSUE TITLE
bugprone-narrowing-conversions false positive for contional expression (in C)

### ISSUE THREAD

**Issue by AndreyG** [llvm/llvm-project#139467]
```c
char test(int cond, char c) {
	char ret = cond > 0 ? ':' : c; // warning: narrowing conversion from 'int' to signed type 'char' is implementation-defined
	return ret;
}
```
https://godbolt.org/z/6P8Yaqn4o

**Comment by pinskia** [comment]
Well in C the type of a character literal is int ...
So `typeof(cond > 0 ? ':' : c)` will be int.

**Comment by AndreyG** [comment]
> So `typeof(cond > 0 ? ':' : c)` will be int.

Yes, but because we know that both arguments of the conditional expression could be converted to `char` without narrowing, then the whole expression could be converted to `char` without narrowing. Moreover the check of conditional expression is already implemented in terms of the separate checks of arguments: https://github.com/llvm/llvm-project/blob/7500cead4e000d82d62f99f2896ffa94ee08ff7e/clang-tools-extra/clang-tidy/bugprone/NarrowingConversionsCheck.cpp#L557-L560.

**Comment by AndreyG** [comment]
The same issue exists in C++ but only for `short`, not for the `char`: https://godbolt.org/z/jbqzrKGPP.
